### ISSUE TITLE
entgql: generate graphql and relay objects

### ISSUE THREAD

**Issue by a8m** [ent/ent#2351]
As part of the effort started by @cliedeman and @giautm on https://github.com/ent/contrib/pull/235, I propose breaking it into smaller tasks and pushing it upstream gradually. We should start with the most common and requested option - generating GraphQL objects from an Ent schema. I suggest starting with the following but would love to hear if you think we should add more.

### Schema-levlel annotation

- Generate basic GraphQL objects (`type T`), allow configuring its name, directives, implemented interfaces, and its exported fields.
- Generate Relay Connection objects and the rest of the types supported by https://entgo.io/docs/tutorial-todo-gql-paginate/
- Auto-update gqlgen config file

### Project-levlel annotation (entc.go)

- Allow generating types like interfaces or types that are globally used and not specific to a schema. 

## Stage 2:

- Allow configuring mutations (actions) and their inputs on the schema-level and generate both gql and ent code for them.
- Generate queries.

### Implementation ideas

I really like the `entgql.RelayConnection` option defined in https://github.com/ent/contrib/pull/235, but for the types, I think we should make it more customized.

1\. Auto-generate GraphQL and Relay types, but skip fields that are marked with `Skip`.

```go
func (Pet) Annotations() []schema.Annotation {
	return []schema.Annotation{
		entgql.Type()
		entgql.RelayConnection(),
	}
}

func (Pet) Fields() []ent.Field {
	return []ent.Field{
		// ...
		field.String("name").
			Annotation(entgql.Skip()),
	}
}
```

2\. More control on the generated model and fields.

```go
func (Pet) Annotations() []schema.Annotation {
	return []schema.Annotation{
		entgql.Type("GQLType").
			Directive(...).
			Comment(...).
			Implemented(...)
		entgql.RelayConnection(),
	}
}

func (Pet) Fields() []ent.Field {
	return []ent.Field{
		// ...
		field.String("name").
			Annotation(
				entgql.Field("gql_name").
					Comment(...),
			),
	}
}
```

3\. Allow free mapping.

```go
func (Pet) Annotations() []schema.Annotation {
	return []schema.Annotation{
		entgql.Type("GQLType").
			Fields(
				entgql.MapField("gql_field1", "<edge or field name>").
					Comment(...),
				entgql.MapField("gql_field2", "<edge or field name>").
					Directive(...),
				// ...
			),
		entgql.RelayConnection(),
	}
}
```

I have drafts for stage-2, but let's focus on stage-1 first. Thoughts? 

cc @crossworth, @giautm, @cliedeman, @rotemtam, @masseelch, @yonidavidson, @maaft

**Comment by giautm** [comment]
> directive support is for fields only.

Not correct, in the current implement you can apply the directive on any field /or any type.

> but does that mean the the Cursor type and PageInfo type should be controlled on this part?

Agree, the cursor type can be `String` or any named custom scalar type (default to `Cursor`)

**Comment by yonidavidson** [comment]
@giautm, I think your example at https://github.com/ent/ent/issues/2173#issuecomment-1049934454 should be part of gql since it's a common request WDYT?

**Comment by giautm** [comment]
> @giautm, I think your example at https://github.com/ent/ent/issues/2173#issuecomment-1049934454 should be part of gql since it's a common request WDYT?

I thinking away to avoid have both SkipFieldExcept and SkipEdgeExcept in the extension. Because It's cloned code.

**Comment by giautm** [comment]
I'd love to hear from you about the integration between the EntGQL extension and the GQLGen plug-in.

First of all, can we agree that Ent's code generation needs to be executed first? Ent -> GQLGen

When Ent executes the code generation, the EntGQL extension is fetching the mapping information (`gqlgen.yaml`) between the scalar and Go data types. This is causing circular dependencies. I'd suggest dropping this, because the WhereFilter feature will be developed into a plug-in.

So I'm imagining the process going like this:
```
Ent's code generations (+EntGQL Extension) -> GQLGen's Code generation (+EntGQL Plugin, +WhereFilter Plugin).
```

And because, in GQLGen's plug-ins, the order of methods is called: `InjectSourceEarly`, `InjectSourceLate`, `MutateConfig` and `GenerateCode`. So we cannot read the mapping information (`gqlgen.yaml`) as before, but should rely on annotations such as single-source-as-trust.

Options such as `WithScalarMappings`, and `WithMapScalarFunc` should be interchanged and shared between the extension and the plug-in.

**Comment by a8m** [comment]
> First of all, can we agree that Ent's code generation needs to be executed first? Ent -> GQLGen

Yes.

> This is causing circular dependencies ... So we cannot read the mapping information (gqlgen.yaml) as before, but should rely on annotations such as single-source-as-trust.

👍

**Comment by giautm** [comment]
There is one more problem regarding the edge filter support. If we separate the WhereFilters feature into a separate plug-in, how do we connect them?

```GQL
type Todo implements Node {
  id: ID!
  children(..., where: TodoWhereInput): TodoConnection
}
```

**Comment by a8m** [comment]
So, maybe we can live it in entgql as an official feature, but not enable it by default.

**Comment by a8m** [comment]
So, maybe we can leave it in entgql as an official feature, but not enable it by default?

**Comment by giautm** [comment]
> So, maybe we can leave it in entgql as an official feature, but not enable it by default?

Actually, we just need to known the name of the input for reference. But, need some duplicated code to generate the correct name.

**Comment by giautm** [comment]
I want to ask people about API design for relational edges. With a 1-1, or n-1 relationship it's pretty simple:
```gql
type Todo implements Node {
  id: ID!
  category: Category
}
```

However, with 1-n relationships it is a bit different.

For example in case there is no filter, sort and pagination, no Relay specification: (We will not support this API).
```gql
type Category {
  id: ID!
  todos: [Todo]
}
```

With pagination, and sorting:
```gql
type Category {
  id: ID!
  todos(page: Int = 1, limit: Int, orderBy: TodoOrder): [Todo]
}
```

With pagination, sorting, and setting Relay (Need to be enabled with the `WithRelaySpecification` option, `entgql.RelayConnection` on the `todos` edge and `entgql.RelayConnection` on the `Todo` entity)
```gql
type Category implements Node {
  id: ID!
  todos(
    after: Cursor
    first: Int
    before: Cursor
    last: Int
    orderBy: TodoOrder
  ): TodoConnection
}
```

With the filter, we will have the `where` parameter on the edges.
```gql
type Category {
  id: ID!
  todos(page: Int = 1, limit: Int, orderBy: TodoOrder, where: TodoWhereInput): [Todo]
}
```
```gql
type Category implements Node {
  id: ID!
  todos(
    after: Cursor
    first: Int
    before: Cursor
    last: Int
    orderBy: TodoOrder
    where: TodoWhereInput
  ): TodoConnection
}
```

**Comment by a8m** [comment]
I like both suggestions. Just note that the code block above can be simplified as follows:

```go
const (
	SkipFlagEnum SkipFlag = 1 << iota
	SkipFlagOrder
	SkipFlagInput
	SkipFlagWhere

	...
)
```

**Comment by yonidavidson** [comment]
Great work @giautm,
I do think we have a challenge on explaining to a user why specific cases will not work (invalid) or won't be generated (not supported) . Considering that gql will help graphQL users to adopt best practices by guiding them in the schema creation stage. I think that very good error/warning messages and good docs can do that.

**Comment by cliedeman** [comment]
Hi everyone. I don't currently work on any ent + gqlgen proejcts so I wanted to share all my thoughts and ideas I had while I was working on this to hopefully help someone.

## Leveraging gqlgen schema config

You don't have to customize the schema exclusively in ent


Generated by ent (you can have a single type with extends)
```graphql
extend type Person {
  firstname
  lastname 
  ...
}
```

User written 
```graphql
extend type Person implements IInterface { # add interfaces
  fullname # will generate a resolver user can implement
   ...
   # any extra frields 
}
```

# Skip Annotation

My idea around `Skip` was to eventually add a complementary `Include` anotation like Relay
The idea being that the user can opt into either include by default or exclude by default and use the relevant annotation.
I personally prefer to stipulate and limit the exposed fields and also not accidentally expose any new fields.

# goModel directive - https://github.com/ent/contrib/pull/247

I never considered this but this will expose some information into your graphql schema. 

```graphql
type User implements Node @goModel(model: "entgo.io/contrib/entgql/internal/todoplugin/ent.User") {
	...
}
```

Probably not a problem but it should probably be discussed

# GQLGen: Plugin Signatures

https://github.com/99designs/gqlgen/discussions/1994
https://github.com/99designs/gqlgen/discussions/1995

Some changes I proposed to gqlgen to make writing plugins easier for us, mostly because we make heavy use of InjectSourceEarly

# GQLGen: Only generate resolver

https://github.com/99designs/gqlgen/discussions/1993

By default `goField(forceResolve: true)` will generate duplicate models which makes usage difficult. 

# Future

## DataLoaders

I created a data loader generator at some point: https://github.com/cliedeman/entdataloader.
I think that this is a key piece for graphql integration. 

## Unions (far future)

I find that some sort of table inheritance (would have to be implemented in ent) that exposes easily into graphql is a (in my experience) common use case

Ciaran

**Comment by giautm** [comment]
> # goModel directive - [ent/contrib#247](https://github.com/ent/contrib/pull/247)
> I never considered this but this will expose some information into your graphql schema.
> 
> ```graphql
> type User implements Node @goModel(model: "entgo.io/contrib/entgql/internal/todoplugin/ent.User") {
> 	...
> }
> ```
> 
> Probably not a problem but it should probably be discussed

It cannot expose via GQL introspect, but it can be expose via `query { _service { sdl } }` query if you using Federation, damm.

**Comment by mojerro** [comment]
I have a problem, can we auto-generate `.graphql` file by following the schema files? Just like what [Hasura](https://github.com/hasura/graphql-engine)  did. Now we should write both schema and Graphql files, I think it is the duplicate work.

**Comment by giautm** [comment]
Hello @mojerro, this issue is for that support. We will generate `.graph` file from the Ent's schema.

**Comment by mojerro** [comment]
> Hello @mojerro, this issue is for that support. We will generate `.graph` file from the Ent's schema.

Oh, that's great. I came from another issue and didn't read the dialog. Sorry 😢  I'll read it carefully

**Comment by mojerro** [comment]
I just see that the work has nearly been done. Is there something else that needs to do or have a  test?

**Comment by mojerro** [comment]
I just see that most of the PRs have been merged. Is there something else that needs to do or have a  test?

**Comment by crossworth** [comment]
I think there is a way to allow easy customization of the `WhereFilters` without templates, right now with [this PR](https://github.com/ent/contrib/pull/275) is possible to add custom fields to the generated `.graphql` file, but to add the query predicate we must use templates, in my case sometimes I just want to add a simple logic for a column that don't really exists on the database.

Like if the current user is owner of the resource.

**Here's the idea for the following user case: Adding a 'isOwner' predicate.** 

Extend the WhereInput generated by ent on another GraphQL file.
```graphql
extend input ChannelWhereInput {
  isOwner: Boolean
}
```
After running GQLGen this will generate the following code:

```go
func (r *channelWhereInputResolver) IsOwner(ctx context.Context, obj *ent.ChannelWhereInput, data *bool) error {
	panic(fmt.Errorf("not implemented"))
}
```

If the generated WhereFilter struct had a custom predicate field I could do something like:

```go
func (r *channelWhereInputResolver) IsOwner(ctx context.Context, obj *ent.ChannelWhereInput, data *bool) error {
	if data == nil {
		return nil
	}

	u, err := r.UserFromContext(ctx)
	if err != nil {
		return err
	}

	if *data {
		obj.Predicates = append(obj.Predicates, channel.UserID(u.ID))
	} else {
		obj.Predicates = append(obj.Predicates, channel.UserIDNEQ(u.ID))
	}
	
	return nil
}
```

Example implementation:
(I'm not sure if the and/or/not are correct and I'm sure the logic for merging the custom predicates could be improved)

```go
// ChannelWhereInput represents a where input for filtering Channel queries.
type ChannelWhereInput struct {
	Predicates []predicate.Channel  `json:"-"`  // <-- Add this field
	Not        *ChannelWhereInput   `json:"not,omitempty"`
	Or         []*ChannelWhereInput `json:"or,omitempty"`
	And        []*ChannelWhereInput `json:"and,omitempty"`
	// ... other fields
}

// P returns a predicate for filtering channels.
// An error is returned if the input is empty or invalid.
func (i *ChannelWhereInput) P() (predicate.Channel, error) {
	var predicates []predicate.Channel
	if i.Not != nil {
		p, err := i.Not.P()
		if err != nil {
			return nil, err
		}
		predicates = append(predicates, channel.Not(p))
                
		// added this block
		for _, p := range i.Predicates {
			predicates = append(predicates, channel.Not(p))
		}
	}
	switch n := len(i.Or); {
	case n == 1:
		p, err := i.Or[0].P()
		if err != nil {
			return nil, err
		}

		// added this block
		if len(i.Predicates) == 0 {
			predicates = append(predicates, p)
		} else {
			predicates = append(predicates, channel.Or(append([]predicate.Channel{p}, i.Predicates...)...))
		}
	case n > 1:
		or := make([]predicate.Channel, 0, n)
		for _, w := range i.Or {
			p, err := w.P()
			if err != nil {
				return nil, err
			}
			or = append(or, p)
		}
		predicates = append(predicates, channel.Or(or...))
		predicates = append(predicates, channel.Or(i.Predicates...)) // <-- added
	}
	switch n := len(i.And); {
	case n == 1:
		p, err := i.And[0].P()
		if err != nil {
			return nil, err
		}

		// added this block
		if len(i.Predicates) == 0 {
			predicates = append(predicates, p)
		} else {
			predicates = append(predicates, channel.And(append([]predicate.Channel{p}, i.Predicates...)...))
		}
	case n > 1:
		and := make([]predicate.Channel, 0, n)
		for _, w := range i.And {
			p, err := w.P()
			if err != nil {
				return nil, err
			}
			and = append(and, p)
		}
		predicates = append(predicates, channel.And(and...))
		predicates = append(predicates, channel.And(i.Predicates...)) // <-- added
	}
	predicates = append(predicates, i.Predicates...) // <-- added
	if i.IsDefaultChannel != nil {
		predicates = append(predicates, channel.IsDefaultChannelEQ(*i.IsDefaultChannel))
	}
	if i.IsDefaultChannelNEQ != nil {
		predicates = append(predicates, channel.IsDefaultChannelNEQ(*i.IsDefaultChannelNEQ))
	}
	if i.Name != nil {
		predicates = append(predicates, channel.NameEQ(*i.Name))
	}
	if i.NameNEQ != nil {
		predicates = append(predicates, channel.NameNEQ(*i.NameNEQ))
	}
	if len(i.NameIn) > 0 {
		predicates = append(predicates, channel.NameIn(i.NameIn...))
	}
	if len(i.NameNotIn) > 0 {
		predicates = append(predicates, channel.NameNotIn(i.NameNotIn...))
	}
	if i.NameGT != nil {
		predicates = append(predicates, channel.NameGT(*i.NameGT))
	}
	if i.NameGTE != nil {
		predicates = append(predicates, channel.NameGTE(*i.NameGTE))
	}
	if i.NameLT != nil {
		predicates = append(predicates, channel.NameLT(*i.NameLT))
	}
	if i.NameLTE != nil {
		predicates = append(predicates, channel.NameLTE(*i.NameLTE))
	}
	if i.NameContains != nil {
		predicates = append(predicates, channel.NameContains(*i.NameContains))
	}
	if i.NameHasPrefix != nil {
		predicates = append(predicates, channel.NameHasPrefix(*i.NameHasPrefix))
	}
	if i.NameHasSuffix != nil {
		predicates = append(predicates, channel.NameHasSuffix(*i.NameHasSuffix))
	}
	if i.NameEqualFold != nil {
		predicates = append(predicates, channel.NameEqualFold(*i.NameEqualFold))
	}
	if i.NameContainsFold != nil {
		predicates = append(predicates, channel.NameContainsFold(*i.NameContainsFold))
	}

	switch len(predicates) {
	case 0:
		return nil, fmt.Errorf("empty predicate ChannelWhereInput")
	case 1:
		return predicates[0], nil
	default:
		return channel.And(predicates...), nil
	}
}
```

I have made a few tests and this works and is very easy to work with.

What do you guys think?

cc @a8m @giautm

**Comment by mojerro** [comment]
I have a question, how should I deal with the enum type with custom GoType.

```go
package schema

func (Profile) Fields() []ent.Field {
	return []ent.Field{
		field.Enum("gender").GoType(property.Gender(property.Unknown)),
}
```

```go
package property

type Gender string

const (
	Unknown Gender = "UNKNOWN"
	Female  Gender = "FEMALE"
	Male    Gender = "MALE"
	Other   Gender = "OTHER"
)

// Values provides list valid values for Enum.
func (Gender) Values() (kinds []string) {
	for _, s := range []Gender{Female, Male, Other, Unknown} {
		kinds = append(kinds, string(s))
	}
	return
}
```
Here is the error, which indicates that the custom GoType needs implements the interface. I'm not sure can entc do it automately.
```
validation failed: packages.Load: /home/mojerro/project/mo-auth-center/internal/data/ent/cache/cache.go:143:24: cannot use property.CacheType("") (constant "" of type property.CacheType) as graphql.Marshaler value in variable declaration: property.CacheType does not implement graphql.Marshaler (missing method MarshalGQL)
/home/mojerro/project/mo-auth-center/internal/data/ent/cache/cache.go:145:26: cannot use (*property.CacheType)(nil) (value of type *property.CacheType) as graphql.Unmarshaler value in variable declaration: *property.CacheType does not implement graphql.Unmarshaler (missing method UnmarshalGQL)
```

While the following code runs well:

```go
package schema

func (Profile) Fields() []ent.Field {
	return []ent.Field{
		field.Enum("gender").
                    // GoType(property.Gender(property.Unknown))
                    Values("UNKNOWN", "FEMALE", "MALE").,
}
```

**Comment by giautm** [comment]
Ent also generates MarshalGQL/UnmarshalGQL for the enum, when you use custom GoType. You must define it by yourseft, the easy way is copying the implement from Ent.
```go
// MarshalGQL implements graphql.Marshaler interface.
func (_g Gender) MarshalGQL(w io.Writer) {
	io.WriteString(w, strconv.Quote(_g.String()))
}

// UnmarshalGQL implements graphql.Unmarshaler interface.
func (_g *Gender) UnmarshalGQL(val interface{}) error {
	str, ok := val.(string)
	if !ok {
		return fmt.Errorf("enum %T must be a string", val)
	}
	*_g = Type(str)

	switch *_g {
	case Unknown, Female, Male, Other:
		return nil
	default:
		return fmt.Errorf("%s is not a valid Gender", str)
	}
}
```

**Comment by a8m** [comment]
The feature was added to `ent/contrib`. Please use `go get entgo.io/contrib@master` to use and see https://entgo.io/docs/tutorial-todo-gql-schema-generator for more info.

Thanks, @crossworth, @yonidavidson, @cliedeman, and special thanks for @giautm ❤️

**Comment by alexander-ushakov-93** [comment]
Hasura resolver-free design based on compiler, generating SQL code, can be more performant replacement of gqlgen. Can you, @a8m, suggest - have you any plans to integrate Hasura as entgo extension?
